File: core/src/muscle_group.cpp
#include "workoutlog/muscle_group.hpp"

namespace workoutlog {

std::string label(MuscleGroup g) {
    switch (g) {
        case MuscleGroup::chest: return "Chest";
        case MuscleGroup::back: return "Back";
        case MuscleGroup::shoulders: return "Shoulders";
        case MuscleGroup::arms: return "Arms";
        case MuscleGroup::legs: return "Legs";
        case MuscleGroup::core: return "Core";
    }
    return {};
}

std::string hex(MuscleGroup g) {
    switch (g) {
        case MuscleGroup::chest: return "#d1495b";
        case MuscleGroup::back: return "#00798c";
        case MuscleGroup::shoulders: return "#edae49";
        case MuscleGroup::arms: return "#8e5ea2";
        case MuscleGroup::legs: return "#30638e";
        case MuscleGroup::core: return "#58a65c";
    }
    return {};
}

std::optional<MuscleGroup> muscle_group_of(std::string_view muscle) {
    if (muscle == "chest") return MuscleGroup::chest;
    if (muscle == "lats" || muscle == "rhomboids" || muscle == "traps" || muscle == "spinal_erectors")
        return MuscleGroup::back;
    if (muscle == "front_delts" || muscle == "side_delts" || muscle == "rear_delts")
        return MuscleGroup::shoulders;
    if (muscle == "biceps" || muscle == "triceps" || muscle == "forearms") return MuscleGroup::arms;
    if (muscle == "quads" || muscle == "hamstrings" || muscle == "glutes" || muscle == "calves" ||
        muscle == "adductors" || muscle == "hip_flexors")
        return MuscleGroup::legs;
    if (muscle == "abs" || muscle == "obliques") return MuscleGroup::core;
    return std::nullopt;
}

std::optional<MuscleGroup> dominant(const std::map<std::string, double>& scores) {
    std::map<MuscleGroup, double> by_group;
    for (const auto& [muscle, score] : scores) {
        auto g = muscle_group_of(muscle);
        if (!g) continue;
        by_group[*g] += score;
    }
    if (by_group.empty()) return std::nullopt;

    std::optional<MuscleGroup> best;
    double best_score = 0;
    for (MuscleGroup g : kMuscleGroupOrder) {
        auto it = by_group.find(g);
        if (it == by_group.end()) continue;
        if (!best || it->second > best_score) {
            best = g;
            best_score = it->second;
        }
    }
    return best;
}

} // namespace workoutlog

File: crates/workout-log-core/src/analytics/muscle_group.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MuscleGroup {
    Chest,
    Back,
    Shoulders,
    Arms,
    Legs,
    Core,
}

impl MuscleGroup {
    /// Declaration order, which is also the tie-break order for [`dominant`].
    pub const ALL: [Self; 6] = [
        Self::Chest,
        Self::Back,
        Self::Shoulders,
        Self::Arms,
        Self::Legs,
        Self::Core,
    ];

    pub fn hex(self) -> &'static str {
        match self {
            Self::Chest => "#d1495b",
            Self::Back => "#00798c",
            Self::Shoulders => "#edae49",
            Self::Arms => "#8e5ea2",
            Self::Legs => "#30638e",
            Self::Core => "#58a65c",
        }
    }

    pub fn name(self) -> &'static str {
        match self {
            Self::Chest => "chest",
            Self::Back => "back",
            Self::Shoulders => "shoulders",
            Self::Arms => "arms",
            Self::Legs => "legs",
            Self::Core => "core",
        }
    }

    pub fn label(self) -> &'static str {
        match self {
            Self::Chest => "Chest",
            Self::Back => "Back",
            Self::Shoulders => "Shoulders",
            Self::Arms => "Arms",
            Self::Legs => "Legs",
            Self::Core => "Core",
        }
    }

    pub fn of(muscle: &str) -> Option<Self> {
        Some(match muscle {
            "chest" => Self::Chest,
            "lats" | "rhomboids" | "traps" | "spinal_erectors" => Self::Back,
            "front_delts" | "side_delts" | "rear_delts" => Self::Shoulders,
            "biceps" | "triceps" | "forearms" => Self::Arms,
            "quads" | "hamstrings" | "glutes" | "calves" | "adductors" | "hip_flexors" => {
                Self::Legs
            }
            "abs" | "obliques" => Self::Core,
            _ => return None,
        })
    }

    /// Ties go to the earlier group in declaration order; None when no muscle maps.
    pub fn dominant(scores: &BTreeMap<String, f64>) -> Option<Self> {
        let mut totals: BTreeMap<Self, f64> = BTreeMap::new();
        for (muscle, score) in scores {
            if let Some(group) = Self::of(muscle) {
                *totals.entry(group).or_insert(0.0) += score;
            }
        }
        let mut best: Option<(Self, f64)> = None;
        for group in Self::ALL {
            if let Some(&total) = totals.get(&group)
                && best.is_none_or(|(_, high)| total > high)
            {
                best = Some((group, total));
            }
        }
        best.map(|(group, _)| group)
    }
}
